django/thunderstore/api/cyberstorm/views/package_preview.py

```diff
@@ -0,0 +1,288 @@
+import math
+from itertools import chain
+from typing import OrderedDict
+
+from django.contrib.postgres.aggregates import ArrayAgg
+from django.core.paginator import EmptyPage, Page
+from django.db.models import (
+    Case,
+    CharField,
+    Count,
+    Lookup,
+    Q,
+    QuerySet,
+    Sum,
+    Value,
+    When,
+)
+from django.http import HttpRequest, HttpResponse
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import status
+from rest_framework.exceptions import ParseError
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from thunderstore.api.cyberstorm.serializers.package import (
+    CommunityPackageListCyberstormSerializer,
+    PackageCategoryCyberstormSerializer,
+    PackageSearchQueryParameterCyberstormSerializer,
+)
+from thunderstore.cache.enums import CacheBustCondition
+from thunderstore.cache.pagination import CachedPaginator
+from thunderstore.community.models import PackageListingSection
+from thunderstore.community.models.package_category import PackageCategory
+from thunderstore.community.models.package_listing import PackageListing
+
+
+class LowerContainedBy(Lookup):
+    lookup_name = "icontained_by"
+
+    def as_sql(self, compiler, connection):
+        lhs, lhs_params = self.process_lhs(compiler, connection)
+        rhs, rhs_params = self.process_rhs(compiler, connection)
+        params = lhs_params + rhs_params
+        return f"LOWER({rhs}) LIKE '%%' || LOWER({lhs}) || '%%'", params
+
+
+CharField.register_lookup(LowerContainedBy)
```

Comment:
Why is this necessary? Not saying it can't be used, but that I'd be surprised there's no way to implement what this is currently used for without relying on custom lookups, as there hasn't been a need to use one until now and we aren't doing anything special in this view.

Reply:
There wasn't anything that would do the same thing on DB level, so I opted to using this one. Probably premature optimization, but your call if to toss it or not?

Reply:
I'm more wondering what part of the queries being implemented actually require this lookup. I mean I did read the code where it was used on a high level, but what does this actually allow us to do that we've not been able to do elsewhere in the codebase without resorting to a custom lookup?

If the performance impact is significant, it might be worth it, but again the complexity cost is pretty high here so I'd default to not including it unless necessary.

Reply:
So because the review statuses are strings and not enums(for example) and we want to check if the status is either `unreviewed` or `approved`; we have to do icontains, but the other way around.

e.g. take the db value and check if it's contained in the string `unreviewed approved`

You can find this check in the first `Case` lookup of `get_queryset`. (or wherever the `icontained_by` is used, if the code has changed)

We could change the statuses to be enums
rejected=1
unreviewed=2
approved=3
and then just do a comparison of something like
`review_status__gte=1`

But I have no preference, let me know which way you think is better 🤷

Reply:
What's stopping you from using exact comparisons against a set of valid options (i.e. `field__in=("unreviewed", "approved")`? Why do you need to use a string contains check?

The values might be strings but they're still limited to a fixed set of options

Reply:
Huh, never crossed my mind, that I could use `__in`, I'll try that out 🤷

Reply:
Which brings us to the necessity of using Case/When in the first place. Regardless, even if you feel the need to rely on Case/When, icontains is one of the worse checks you can use for this use case and should not be used. Simply put, it's the wrong tool for the job.